homeassistant/components/sma/config_flow.py

```diff
@@ -0,0 +1,215 @@
+
+    protocol = "https" if data[CONF_SSL] else "http"
+    url = f"{protocol}://{data[CONF_HOST]}"
+
+    sma = pysma.SMA(session, url, data[CONF_PASSWORD], group=data[CONF_GROUP])
+
+    if await sma.new_session() is False:
+        raise InvalidAuth
+
+    device_info = await sma.device_info()
+
+    if not device_info:
+        raise CannotRetrieveDeviceInfo
+
+    await sma.close_session()
+    return device_info
+
+
+class SmaConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for SMA."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+
+    def __init__(self):
+        """Initialize."""
+        self._data = {
+            CONF_HOST: vol.UNDEFINED,
+            CONF_SSL: False,
+            CONF_VERIFY_SSL: True,
+            CONF_GROUP: GROUPS[0],
+            CONF_PASSWORD: vol.UNDEFINED,
+            CONF_SENSORS: [],
+            CONF_CUSTOM: {},
+            DEVICE_INFO: {},
+        }
+
+    async def async_step_user(self, user_input=None):
+        """First step in config flow."""
+        errors = {}
+        if user_input is not None:
+            self._data[CONF_HOST] = user_input[CONF_HOST]
+            self._data[CONF_SSL] = user_input[CONF_SSL]
+            self._data[CONF_VERIFY_SSL] = user_input[CONF_VERIFY_SSL]
+            self._data[CONF_GROUP] = user_input[CONF_GROUP]
+            self._data[CONF_PASSWORD] = user_input[CONF_PASSWORD]
+
+            try:
+                self._data[DEVICE_INFO] = await validate_input(self.hass, user_input)
+            except aiohttp.ClientError:
+                errors["base"] = "cannot_connect"
+            except InvalidAuth:
+                errors["base"] = "invalid_auth"
+            except CannotRetrieveDeviceInfo:
+                errors["base"] = "cannot_retrieve_device_info"
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
+
+            if not errors:
+                await self.async_set_unique_id(self._data[DEVICE_INFO]["serial"])
+                self._abort_if_unique_id_configured()
+                return await self.async_step_sensors()
+
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(CONF_HOST, default=self._data[CONF_HOST]): cv.string,
+                    vol.Optional(CONF_SSL, default=self._data[CONF_SSL]): cv.boolean,
+                    vol.Optional(
+                        CONF_VERIFY_SSL, default=self._data[CONF_VERIFY_SSL]
+                    ): cv.boolean,
+                    vol.Optional(CONF_GROUP, default=self._data[CONF_GROUP]): vol.In(
+                        GROUPS
+                    ),
+                    vol.Required(
+                        CONF_PASSWORD, default=self._data[CONF_PASSWORD]
```

Comment:
We shouldn't send the password back to the UI once we have it.

Reply:
Good catch. You're right!